Detach mod customisation area from the footer and replace with an overlay panel display

frenzibyte opened the pull request:
- [x] Depends on https://github.com/ppy/osu-framework/pull/6325
- [x] Depends on framework bump

Prerequisite for integrating the new `ScreenFooter` component with the existing mod select overlay screens (alongside other sheared overlays).

The current mod customisation area is ingrained underneath the footer. When replacing the overlay's local footer with `ScreenFooter`, extra work needs to be put into `ScreenFooter` to work with the mod customisation area as is (especially since `ScreenFooter` is global).

Instead, this PR replaces the current mod customisation area with an overlay panel view instead, detaching it from the footer completely. Inspired by the [reference design](https://www.figma.com/design/NkdREXr6wFkA99oLw8PnBv/Client%2FMod-Selection?node-id=2292-373&t=aAyYYyngDawZb8AS-4).

Preview:

https://github.com/ppy/osu/assets/22781491/76675355-f191-4576-a8f7-43a7f4c53f9b

frenzibyte committed:
Add different display for mod customisation

frenzibyte committed:
Use new display on mod select overlay and remove old display

frenzibyte committed:
Update test coverage

frenzibyte committed:
Remove stale comment

bdach commented:
Sorry but on a literal five minute inspection this is too broken in its current state.

- If tab is pressed when this is expanded and random's seed text box is focused, focus will be taken away from the random textbox which arguably is indesirable.
- Moreover, pressing backspace in the random seed text box instead causes all mods to be deselected. What's worse, re-selecting random after doing that seems to somehow break the textbox so that it can't accept input anymore.

https://github.com/ppy/osu/assets/20418176/22d08310-d216-4a08-8d3d-334c9998c7cb

- 1px masking artifacts on the dropdown when it is collapsed:

![1719209089](https://github.com/ppy/osu/assets/20418176/d3ed7649-1369-4d7d-9863-2b399a7a9599)

- The expandable attribute display can overlap the dropdown and lead to terrible behaviour like having to drag a slider that's obscured by something else:

https://github.com/ppy/osu/assets/20418176/f1682bad-d7b6-4ec3-90c7-732b0a87c4fc

- Dropdowns work terribly with this because you can get into situations wherein you click the header but can't see the actual dropdown choices:

https://github.com/ppy/osu/assets/20418176/07615c4d-8b79-49ad-be41-1d1f194d280a

bdach reviewed (changes requested):
see above

frenzibyte commented:
> Sorry but on a literal five minute inspection this is too broken in its current state.
> 
> * If tab is pressed when this is expanded and random's seed text box is focused, focus will be taken away from the random textbox which arguably is indesirable.
> * Moreover, pressing backspace in the random seed text box instead causes all mods to be deselected. What's worse, re-selecting random after doing that seems to somehow break the textbox so that it can't accept input anymore.

Whoops, I forgot mod settings based on textboxes exist 😅



> * 1px masking artifacts on the dropdown when it is collapsed:

Classic, not sure how I didn't notice that though. I'm gonna have to dig through the existing workarounds we have for this and hope it works.

> * The expandable attribute display can overlap the dropdown and lead to terrible behaviour like having to drag a slider that's obscured by something else:

Decreasing the maximum height should be enough to fix this.

> * Dropdowns work terribly with this because you can get into situations wherein you click the header but can't see the actual dropdown choices:

Hmm, this is a tough one. I'll try to add special logic to `OsuDropdown`s so that it searches for a nearest `ScrollContainer` parent and make sure it's scrolled into view of the dropdown (probably isolated to mod setting dropdowns).

frenzibyte committed:
Fix customisation panel conflicting with beatmap attributes when collapsed

frenzibyte committed:
Add extra margin to avoid 1px artifacts

frenzibyte committed:
Block certain operations when customisation panel is open

Normally I would just block keyboard input from going past `ModCustomisationPanel`, but it's a little complicated here since I'm dealing with global action key binding presses, and I also still want actions like `GlobalAction.Back` to get past the customisation panel even if it's expanded.

frenzibyte committed:
Scroll mod setting dropdown into view when open

frenzibyte committed:
Kill focus from components when clicking on an empty space in the panel

frenzibyte commented:
All concerns should be properly resolved now ~~(hope you spend more than five minutes this time >.>)~~.

bdach commented:
Better, but I still see various weirdness:

If you open the customisation dropdown first and then try to add a mod preset, the customisation dropdown will close, but if you try to add a mod preset _and then_ open the customisation dropdown, it will not close:

https://github.com/ppy/osu/assets/20418176/61784e34-1697-4318-b75a-60dcf6c2fb2a

I'm honestly not sure what the correct behaviour is here but the order of open should not matter.

---

If you click away from the customisation menu by clicking on a column background, the search box will regain focus on mouse down and then lose it again on mouse up:

https://github.com/ppy/osu/assets/20418176/f351a4fa-0118-41e5-af6a-9360fd5ccee8

---

Mod hotkeys work with the customisation menu open. The first mod hotkey press will change active mods and as such close the customisation menu, at which point the search will regain focus, and the second mod hotkey press will start *inputting search terms* into the search box instead:

https://github.com/ppy/osu/assets/20418176/7a26c87d-0f90-4783-8d88-a4886d25d047

frenzibyte committed:
Reorder container layout for popovers to recognize clicks on customisation panel

Basically moves `PopoverContainer` to cover both the columns and the customisation panel, so that if the customisation panel is clicked on, the popover container will notice that and hide popovers like the mod preset popover.

frenzibyte committed:
Improve UX & input handling when customisation panel is open

frenzibyte commented:
> If you open the customisation dropdown first and then try to add a mod preset, the customisation dropdown will close, but if you try to add a mod preset and then open the customisation dropdown, it will not close:
> I'm honestly not sure what the correct behaviour is here but the order of open should not matter (either both should be allowed to be open at once, or neither, regardless of order).

It's better for the mod preset to be closed in this case, this is supposed to just work but the popover container is placed at a level where it cannot notice clicks on the customisation panel header.

> If you click away from the customisation menu by clicking on a column background, the search box will regain focus on mouse down and then lose it again on mouse up:

This is a complicated issue caused by the customisation panel hiding on a mouse-down instead of a click. I've made it as such to match with popovers (dismiss on mouse down), but continuing to work with that behaviour while trying to fix issues with search box focus turned out to be a nightmare, so I'll just change the behaviour to hide on a click instead (which entails dimming the mod columns because the user will no longer be able to select a mod with a single click while the customisation panel is open).

> Mod hotkeys work with the customisation menu open. The first mod hotkey press will change active mods and as such close the customisation menu, at which point the search will regain focus, and the second mod hotkey press will start inputting search terms into the search box instead:

I'm gonna have to change the customisation panel to be a complete input-blocking overlay container for this one and call it a day.

frenzibyte committed:
Add test coverage for the issues that were pointed out recently

frenzibyte committed:
Fix test failure

Joehuu commented on osu.Game/Overlays/Mods/ModCustomisationPanel.cs:
```diff
@@ -0,0 +1,198 @@
+        protected override void Update()
+        {
+            base.Update();
+            scrollContainer.Height = Math.Min(scrollContainer.AvailableContent, DrawHeight - header_height);
```
Seems a bit buggy when changing ui scale:

https://github.com/ppy/osu/assets/35318437/c667c70a-2d6d-4218-9baf-b1c727b8a126

Joehuu commented on osu.Game/Overlays/Mods/ModSelectOverlay.cs:
```diff
@@ -506,41 +493,32 @@ private void updateCustomisation()
-
-                customisationVisible.Disabled = true;
+                customisationPanel.Expanded.Value = false;
+                customisationPanel.Hide();
```
Wondering if this should just always show but in a disabled state (like old behavior with the footer button).

Feels weird to me to have it hidden as a first impression (like the feature disappeared / isn't there). And when enabling Difficulty Adjust for the first time, I don't have a sense of where the panel will appear/expand.

Can also clarify further with a disabled tooltip saying `Select at least one mod that is customisable.` or something.

frenzibyte replied:
Applied in https://github.com/ppy/osu/pull/28553/commits/1d94c96a8efef24d5ae54aa7159f6e7a615a50a8

frenzibyte committed:
Display customisation header in disabled state when no customisable mod selected

frenzibyte committed:
Merge branch 'master' into mod-select-customisation-panel

bdach committed:
Fix test

bdach commented on osu.Game/Configuration/SettingSourceAttribute.cs:
```diff
@@ -274,8 +276,37 @@ private partial class ModSettingsEnumDropdown<T> : SettingsEnumDropdown<T>
+                                float target = scroll.GetChildPosInContent(this, new Vector2(0, newSize.Y + padding));
+
+                                if (target > scroll.Current + scroll.DisplayableContent)
+                                    scroll.ScrollTo(target - scroll.DisplayableContent);
```
While we're on this I would like to say that I find that the current implementation of this, which smushes this behaviour into `SettingSourceAttribute`, is very very very *very **very*** bad. It is a situational interaction which got put into a component that is *not* mod setting specific anymore.

To that end I would propose the following diff over the current code, which addresses both the code quality and the behaviour concerns simultaneously:

```diff
diff --git a/osu.Game/Configuration/SettingSourceAttribute.cs b/osu.Game/Configuration/SettingSourceAttribute.cs
index f87c8de0cb..1e425c88a6 100644
--- a/osu.Game/Configuration/SettingSourceAttribute.cs
+++ b/osu.Game/Configuration/SettingSourceAttribute.cs
@@ -13,10 +13,8 @@
 using osu.Framework.Extensions.TypeExtensions;
 using osu.Framework.Graphics;
 using osu.Framework.Localisation;
-using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Overlays.Settings;
-using osuTK;
 
 namespace osu.Game.Configuration
 {
@@ -276,37 +274,8 @@ private partial class ModSettingsEnumDropdown<T> : SettingsEnumDropdown<T>
 
             private partial class ModDropdownControl : DropdownControl
             {
-                protected override DropdownMenu CreateMenu() => new ModDropdownMenu();
-
-                private partial class ModDropdownMenu : OsuDropdownMenu
-                {
-                    public ModDropdownMenu()
-                    {
-                        // Set menu's max height low enough to workaround nested scroll issues (see https://github.com/ppy/osu-framework/issues/4536).
-                        MaxHeight = 100;
-                    }
-
-                    protected override void UpdateSize(Vector2 newSize)
-                    {
-                        base.UpdateSize(newSize);
-
-                        // todo: probably move this to OsuDropdown so that settings overlay can benefit from this as well.
-                        if (newSize.Y > 0)
-                        {
-                            var scroll = this.FindClosestParent<OsuScrollContainer>();
-
-                            if (scroll != null)
-                            {
-                                const float padding = 15;
-
-                                float target = scroll.GetChildPosInContent(this, new Vector2(0, newSize.Y + padding));
-
-                                if (target > scroll.Current + scroll.DisplayableContent)
-                                    scroll.ScrollTo(target - scroll.DisplayableContent);
-                            }
-                        }
-                    }
-                }
+                // Set menu's max height low enough to workaround nested scroll issues (see https://github.com/ppy/osu-framework/issues/4536).
+                protected override DropdownMenu CreateMenu() => base.CreateMenu().With(m => m.MaxHeight = 100);
             }
         }
     }
diff --git a/osu.Game/Overlays/Mods/ModCustomisationPanel.cs b/osu.Game/Overlays/Mods/ModCustomisationPanel.cs
index 240569031c..ef0429f43c 100644
--- a/osu.Game/Overlays/Mods/ModCustomisationPanel.cs
+++ b/osu.Game/Overlays/Mods/ModCustomisationPanel.cs
@@ -10,8 +10,10 @@
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Effects;
 using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.UserInterface;
 using osu.Framework.Input.Bindings;
 using osu.Framework.Input.Events;
+using osu.Framework.Testing;
 using osu.Game.Configuration;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Containers;
@@ -193,6 +195,9 @@ protected override void Update()
         {
             base.Update();
             scrollContainer.Height = Math.Min(scrollContainer.AvailableContent, DrawHeight - header_height);
+
+            if (this.ChildrenOfType<Menu>().FirstOrDefault(m => m.State == MenuState.Open) is Menu openMenu)
+                scrollContainer.ScrollIntoView(openMenu, extraScroll: 30);
         }
 
         private partial class FocusGrabbingContainer : InputBlockingContainer

```

frenzibyte replied:
The entire feature been removed as per a discussion in [Discord](https://discord.com/channels/188630481301012481/188630652340404224/1258271932319531030).

bdach commented on osu.Game/Overlays/Mods/ModCustomisationPanel.cs:
```diff
@@ -0,0 +1,206 @@
+
+        public override bool HandleNonPositionalInput => Expanded.Value;
+
+        protected override bool BlockPositionalInput => true;
```
slightly bothered by this, why is this not `Expanded.Value` too?

frenzibyte replied:
Hopefully https://github.com/ppy/osu/pull/28553/commits/86583898548c83e3211b6c10f653b88382ada954 should explain this. TL;DR this override is redundant and even then `HandlePositionalInput` controls whether the block will be effective or not. `BlockPositionalInput` is not supposed to return a variable anyway, it has to be constant.

bdach reviewed (commented):
seems mostly fine outside of [remaining concern](https://github.com/ppy/osu/pull/28553#discussion_r1663870927)

frenzibyte committed:
Clarify how the panel blocks input

frenzibyte committed:
Remove scroll-dropdown-into-view logic

frenzibyte committed:
Allow scrollbar to overlap content

The content is already padded enough to have the scrollbar sit on top.
Having the content change padding when the scrollbar appears gives an
unpleasent experience (especially when the scrollbar is hidden at first
but the user increases the content's height by clicking on a dropdown or
something)

Joehuu commented on osu.Game/Overlays/Mods/ModCustomisationPanel.cs:
```diff
@@ -0,0 +1,205 @@
+                content.AutoSizeAxes = Axes.Y;
+                content.FadeEdgeEffectTo(0.25f, 120, Easing.OutQuint);
+            }
+            else
+            {
+                content.AutoSizeAxes = Axes.None;
+                content.ResizeHeightTo(header_height, 400, Easing.OutQuint);
+                content.FadeEdgeEffectTo(0f, 400, Easing.OutQuint);
+            }
```
Content never fades out like regular dropdowns and other UI elements.

Joehuu commented on osu.Game/Overlays/Mods/ModCustomisationHeader.cs:
```diff
@@ -0,0 +1,97 @@
+{
+    public partial class ModCustomisationHeader : OsuHoverContainer
+    {
+        public override bool HandlePositionalInput => true;
```
Is this needed? The inherited class `OsuHoverContainer` overrides `On-` input methods so it should already be true.

frenzibyte replied:
Can't remember why I added that there, should be fine to remove.

Joehuu commented on osu.Game/Overlays/Mods/ModCustomisationHeader.cs:
```diff
@@ -0,0 +1,97 @@
+
+            Expanded.BindValueChanged(v =>
+            {
+                icon.RotateTo(v.NewValue ? 180 : 0);
```
Maybe port this to regular dropdowns for consistent UI behavior? In a separate PR of course.

peppy replied:
If we're keeping it, I'd propose this instead:

```diff
diff --git a/osu.Game/Overlays/Mods/ModCustomisationHeader.cs b/osu.Game/Overlays/Mods/ModCustomisationHeader.cs
index 2887b53548..c96f9da00f 100644
--- a/osu.Game/Overlays/Mods/ModCustomisationHeader.cs
+++ b/osu.Game/Overlays/Mods/ModCustomisationHeader.cs
@@ -90,7 +90,7 @@ protected override void LoadComplete()
 
             Expanded.BindValueChanged(v =>
             {
-                icon.RotateTo(v.NewValue ? 180 : 0);
+                icon.ScaleTo(v.NewValue ? new Vector2(1, -1) : Vector2.One, 300, Easing.OutQuint);
             }, true);
         }
     }

```

frenzibyte replied:
Applied in https://github.com/ppy/osu/pull/28553/commits/22f2f8369521abf943db38387ed173d884f6b0ea and opened separate PR to update the other UI components to follow suit (https://github.com/ppy/osu/pull/28762).

Joehuu reviewed (commented):
Looks good to me after this.

frenzibyte committed:
Add fade transition to content during open/close

frenzibyte committed:
Remove unnecessary `HandlePositionalInput` override

frenzibyte committed:
Rescale chevron up and down instead

Joehuu committed:
Fix content overflowing to border

The +2 for the top isn't really needed for the original purpose as content fades out now, but visually, having the header and content spacing looks more correct.

Joehuu commented:
I've fixed the content overflowing to the border:

| Before | After |
| --- | --- |
| <img width="103" alt="Screenshot 2024-07-07 at 11 15 15 PM" src="https://github.com/ppy/osu/assets/35318437/31b1cbdb-f901-417c-8f22-aa1e72b1741e"> | <img width="113" alt="Screenshot 2024-07-07 at 11 13 26 PM" src="https://github.com/ppy/osu/assets/35318437/f21d2409-7c6f-4c74-96d8-4eb400d52443"> |

I've kept the +2 on the top, even if it's not needed anymore (because of the fade out changes), as it looks more visually correct to me (i.e. having the header have some depth). Can just remove if you don't agree with me.

| With margin | Without margin |
| --- | --- |
| <img width="545" alt="Screenshot 2024-07-07 at 11 15 48 PM" src="https://github.com/ppy/osu/assets/35318437/8a1fdd2e-14f0-407e-b9f7-7036293c19f2"> | <img width="471" alt="Screenshot 2024-07-07 at 11 18 13 PM" src="https://github.com/ppy/osu/assets/35318437/d03d5720-d966-41c9-a71b-09ebf7540179"> |

frenzibyte commented on osu.Game/Overlays/Mods/ModCustomisationPanel.cs:
```diff
@@ -161,13 +164,13 @@ private void updateDisplay()
                 content.AutoSizeAxes = Axes.None;
                 content.ResizeHeightTo(header_height, 400, Easing.OutQuint);
-                content.FadeEdgeEffectTo(0f, 400, Easing.OutQuint);
+                content.FadeOut(400, Easing.OutSine);
```
Yup, I personally picked `OutSine` over `OutQuint` because the latter makes the content vanish quite abruptly when the mod customisation panel is too long.

peppy reviewed (approved):
Feeling good to me 👍

bdach reviewed (approved):
people will probably break this further but seems ok enough on a 5 minute check (which is the most i'm willing to give this after this many review rounds)
